Merge 3.0 to 3.1

hmlanigan opened the pull request:
Contains:

* #15099 from hmlanigan/lp1999060-single-charm-multi-app
* #15102 from SimonRichardson/pass-series-in-upgrade-machine-charm-hook
* #15111 from jack-w-shaw/cherry-pick-ck-fix-3.0

Merge conflicts with:
* apiserver/facades/controller/charmdownloader/charmdownloader_test.go 
* state/export_test.go 
* state/migration_import.go 
* state/migration_import_test.go
* tests/suites/ck/ck.sh

SimonRichardson reviewed (changes requested):
Something is wrong with the merge. (see the linters)

wallyworld commented on state/upgrades.go:
```diff
@@ -95,3 +94,62 @@ func applyToAllModelSettings(st *State, change func(*settingsDoc) (bool, error))
+// where the charm has been downloaded, however the origin was not updated
+// with ID and Hash data. Per LP 1999060, usually seen with multi application
+// using same charm bundles.
+func CorrectCharmOriginsMultiAppSingleCharm(pool *StatePool) error {
```
As mentioned in the PR text, this needs to be removed.

jack-w-shaw commented:
I've just merged https://github.com/juju/juju/pull/15111 into 3.0. While you're at it, it would be great if that could be included in this merge

hmlanigan commented:
Added @jack-w-shaw pr to merge, reverted the upgrade steps, redo on a charmdownloader test merge for no macaroons.  Also rebased to the current 3.1

Not sure why I'm getting linter issues on files not changed. Will look after the lint run on current code.

SimonRichardson reviewed (approved):
OK to land if the linters say so.

hmlanigan commented:
When I try to regenerate `core/charm/downloader/mocks/downloader_mocks.go`, no changes are shown. `check-multi-juju` is successful too, the tests run.

hmlanigan commented:
/merge
